articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+```
+
+How to parse a `rmw_qos_profile_t` from the qos file is up to the specific loader.
+
+### Usage in init options and node options
+
+A `rcl_qos_loaded_profiles_t` member could be added in both `rcl_init_options_t` and `rcl_node_options_t`.
+The one in init options will be common to all nodes in the same contexts, and can be overridden by the one in node options.
+
+Optionally, we could also add in `rcl_node_options_t` a boolean to indicate to use the `rcl_qos_loaded_profiles_t` in init options if the profile was not found in the node specific loaded profiles object.
+
+### New ros argument
+
+A new argument will be needed to allow the user to specify which file to load.
+
+```bash
+ros2run <package_name> <exec_name> --ros-args --qos-file /path/to/qos/file
+```
+
+This loading mechanism will be equivalent to setting the `rcl_qos_loaded_profiles_t` member in init options.
+
+## Further extensions
+
+### Loading qos profiles in composable nodes
+
+The interface definition of the [LoadNode](https://github.com/ros2/rcl_interfaces/blob/master/composition_interfaces/srv/LoadNode.srv) service could be extended with a field to specify the desired `QoS` of the loaded node.
+
+In that way, the user loading the node is responsible of passing the desired QoS, instead of the node container.
+The message to be sent could be loaded from a qos profile file.
+
+#### Proposed interfaces
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# EndpointQoS.msg
+
+string name  # This can be either a topic or service name
+uint8 type  # Endpoint type. An enum that can be a publisher/subscription/topic/service
+QoS qos  # Message that defines the qos
+```
+
+```
+# composition_interfaces/srv/LoadNode.srv
+
+# Current request fields
+...
+# New field
+EndpointQoS[] requested_qos_profiles  # new field to allow loading user requested qos profiles
+---
+# Current response definition
+...
+```
+
+A `rcl_qos_loaded_profiles_t` can be constructed from the received message and passed to the node options of the node to be loaded.
+
+TBD: `QoS` message definition.
+
+### Support in launch files
+
+`launch_ros.actions.Node` and `launch_ros.actions.LoadComposableNode` actions could support a qos file argument in their constructors.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
```

Comment:
Now we're starting to creep into system definition territory.

It's somewhere I want to go but let's not go there right now. :stuck_out_tongue: Better to stick to separate QoS files for now I think.

Reply:
> Now we're starting to creep into system definition territory.
> It's somewhere I want to go but let's not go there right now.  Better to stick to separate QoS files for now I think.

Well I think that the proposal is already in the "system definition" territory, besides if we have an unified file or not :joy:.
I think it would be good to consider if this is desired in the future or not, so we can take a better decision now.

Reply:
> I think for now I would err on the side of having a structure that could be merged with the parameter file (i.e. it doesn't conflict in any way) but not try to actually use them together.

Yeah, that's my idea.
I don't aim to merge both file formats now, it would be a lot of work.